lib/pbench/test/unit/server/test_state_tracker.py

```diff
@@ -0,0 +1,200 @@
+        )
+        ds1.add()
+
+        ds2 = Dataset.attach(controller="frodo", name="rover")
+        assert ds2.user == ds1.user
+        assert ds2.controller == ds1.controller
+        assert ds2.name == ds1.name
+        assert ds2.state == States.INDEXING
+        assert ds2.md5 is ds1.md5
+        assert ds2.id is ds1.id
+
+    def test_attach_filename(self):
+        """ Test that we can create a dataset using the full tarball
+        file path.
+        """
+        ds1 = Dataset(
+            user="webb", path="/foo/bilbo/rover.tar.xz", state=States.QUARANTINED
+        )
+        ds1.add()
+
+        ds2 = Dataset.attach(controller="bilbo", name="rover")
+        assert ds2.user == ds1.user
+        assert ds2.controller == ds1.controller
+        assert ds2.name == ds1.name
+        assert ds2.state == States.QUARANTINED
+        assert ds2.md5 is ds1.md5
+        assert ds2.id is ds1.id
+
+    def test_advanced_good(self):
+        """ Test advancing the state of a dataset
+        """
+        ds = Dataset(user="drb", controller="frodo", name="fio")
+        ds.advance(States.UPLOADED)
+
+    def test_advanced_bad(self):
+        """ Test that we can't advance to a state that's not a
+        successor to the initial state.
+        """
+        ds = Dataset(user="drb", controller="frodo", name="fio")
+        with pytest.raises(DatasetBadStateTransition):
+            ds.advance(States.EXPIRED)
+
+    def test_advanced_terminal(self):
+        """ Test that we can't advance from a terminal state
+        """
+        ds = Dataset(user="drb", controller="frodo", name="fio", state=States.EXPIRED)
+        with pytest.raises(DatasetTerminalStateViolation):
+            ds.advance(States.UPLOADING)
+
+    def test_lifecycle(self):
+        """ Advance a dataset through the entire lifecycle using the state
+        transition dict.
+        """
+        ds = Dataset(user="dave", controller="bilbo", name="Fred")
+        assert ds.state == States.UPLOADING
+        beenthere = [ds.state]
+        while ds.state in Dataset.transitions:
+            advances = Dataset.transitions[ds.state]
+            for n in advances:
+                if n not in beenthere:
+                    next = n
+                    break
+            else:
+                break  # avoid infinite reindex loop!
+            beenthere.append(next)
+            ds.advance(next)
+            assert ds.state == next
+        lifecycle = ",".join([s.name for s in beenthere])
+        assert (
+            lifecycle
+            == "UPLOADING,UPLOADED,UNPACKING,UNPACKED,INDEXING,INDEXED,EXPIRING,EXPIRED"
+        )
+
+    def test_metadata(self):
+        """ Various tests on Metadata keys
+        """
+        # See if we can create a metadata row
+        ds = Dataset.create(user="user", controller="controller", name="name")
+        m = Metadata.create(key=Metadata.REINDEX, value="TRUE", dataset=ds)
+        assert m is not None
```

Comment:
Just a fun side note ... I added this after looking at the coverage report and noting that I had never called `Metadata.create()`. Checking coverage is a good thing!